### PR TITLE
Performance improvements for image views.

### LINKED ISSUE #655: Framebuffer takes a lot of time even if it is already cached
I tried to find visual difference between two consecutive commands, so I keep switching between them.
The framebuffer takes non-negligible time to load anyway, and the loading icon eventually show up, which breaks my "human eye diffing algorithm" :-/

### PR DESCRIPTION
**Pull request by pmuetschard**
Fixes #655

### CODE REVIEW

File: gapic/src/main/com/google/gapid/image/ArrayImage.java [hunk 1]
```diff
@@ -302,47 +293,38 @@ protected void convert2D(byte[] src, byte[] dst, byte[] alpha, int stride) {
...
-      out.rewind();
-      return out;
+    public boolean isHDR() {
+      return true;
     }
 
     @Override
-    public boolean isHDR() {
-      return true;
+    public void bin(Histogram.Binner binner) {
```
Comment (ben-clayton) [inline gapic/src/main/com/google/gapid/image/ArrayImage.java]: I'm not sold on this - replacing a very useful 'give me the value of this pixel in a standardized format' with something so heavily tied to the histogram seems like an overall loss. Given that you're calling back to the binner for each pixel component, I can't imagine it's all that much faster.

File: gapic/src/main/com/google/gapid/image/Histogram.java [hunk 2]
```diff
@@ -34,255 +37,271 @@
...
-    for (DoubleBuffer values : allValues.values()) {
-      for (double value : values.array()) {
-        if (!Double.isNaN(value) && !Double.isInfinite(value)) {
-          min = Math.min(min, value);
-          max = Math.max(max, value);
-          average += value;
-          numValues++;
-        }
-      }
+    public double map(double value) {
```
Comment (ben-clayton) [inline gapic/src/main/com/google/gapid/image/Histogram.java]: Why do we have this method declared twice?

File: gapic/src/main/com/google/gapid/image/Histogram.java [hunk 3]
```diff
@@ -34,255 +37,271 @@
...
+    return mapper.range(count);
+  }
 
-    /**
-     * @return the value limited to the min and max values of this range.
-     */
-    public double clamp(double value) {
-      return Math.max(Math.min(value, max), min);
-    }
+  public static class Binner {
```
Comment (ben-clayton) [inline gapic/src/main/com/google/gapid/image/Histogram.java]: Please document all the new code.

File: gapic/src/main/com/google/gapid/image/ArrayImage.java [hunk 4]
```diff
@@ -174,8 +179,10 @@ public Builder flip() {
   public static class RGBA8Image extends ArrayImage {
     private final PixelInfo info;
 
-    public RGBA8Image(int width, int height, int depth, byte[] data) {
-      this(width, height, depth, data, IntPixelInfo.compute(data, true));
+    public RGBA8Image(
+        com.google.gapid.proto.image.Image.ID id, int width, int height, int depth, byte[] data) {
```
Comment (ben-clayton) [inline gapic/src/main/com/google/gapid/image/ArrayImage.java]: I actually attempted adding an ID into all these images and decided against it once I started encountering multiple images (your `getCombinedId`). Given that you're then making up identifiers, things are just gross.
What are your thoughts on the standard `equals` & `hashCode` approach in my PR?
It certainly seems like far less code...

File: gapic/src/main/com/google/gapid/widgets/ImagePanel.java [hunk 5]
```diff
@@ -1056,9 +1057,11 @@ public void init(Renderer renderer) {
     public void update(Renderer renderer, SceneData newData) {
       // Release textures that are no longer in data.
       Set<Image> newSet = Sets.newHashSet(newData.images);
-      for (Map.Entry<Image, Texture> entry : imageToTexture.entrySet()) {
+      for (Iterator<Map.Entry<Image, Texture>> it = imageToTexture.entrySet().iterator(); it.hasNext(); ) {
```
Comment (ben-clayton) [inline gapic/src/main/com/google/gapid/widgets/ImagePanel.java]: While this is marginally faster, I found the implementation in #1346 significantly easier to read.

File: gapic/src/main/com/google/gapid/image/Histogram.java (hunk 3 above)
Comment (pmuetschard) [reply gapic/src/main/com/google/gapid/image/Histogram.java]: Done.

File: gapic/src/main/com/google/gapid/image/ArrayImage.java (hunk 4 above)
Comment (pmuetschard) [reply gapic/src/main/com/google/gapid/image/ArrayImage.java]: Done. I think. It's not really less code, but it certainly is cleaner.

File: gapic/src/main/com/google/gapid/image/ArrayImage.java (hunk 1 above)
Comment (ben-clayton) [reply gapic/src/main/com/google/gapid/image/ArrayImage.java]: Fair enough. I can't argue with your profiling results. I guess if we need to sample a pixel, we can always add that back in again.

**Comment by ben-clayton** [comment]
Thank you for the changes!

**Pull request merged**